napps/mongodb.md

```diff
@@ -0,0 +1,150 @@
+- **Database Models**: It's mainly responsible for defining [pydantic](https://pydantic-docs.helpmanual.io/) database schema models. Although MongoDB is schemaless, these database models provide base document structures just so database relations and operations can be modeled accordingly and easier to evolve and reason about. Also they provide certain validations without a round trip to the database.
+
+### Code Sample
+
+To illustrate a practical example, `upsert_switch` is a method of a controller that's responsible for upserting (update or insert) a switch document into a collection `switches`, notice that `SwitchDoc` is a `pydantic` database model that represents a database document, this method ensures that a `switch_dict` payload that will be parsed accordingly. In this example, the response isn't being used by an endpoint or another client, so it's not mapping to an API model (both models can be the same when applicable):
+
+```
+class SwitchDoc(BaseModel):
+    """Switch DB Document Model."""
+
+    id: str = Field(None, alias="_id")
+    enabled: bool
+    active: bool
+    data_path: Optional[str]
+    hardware: Optional[str]
+    manufacturer: Optional[str]
+    software: Optional[str]
+    connection: Optional[str]
+    ofp_version: Optional[str]
+    serial: Optional[str]
+    metadata: dict = {}
+    interfaces: List[InterfaceSubDoc] = []
+    inserted_at: Optional[datetime]
+    updated_at: Optional[datetime]
+
+class TopoController:
+
+    def upsert_switch(self, dpid: str, switch_dict: dict) -> Optional[dict]:
+        """Update or insert switch."""
+        utc_now = datetime.utcnow()
+        model = SwitchDoc(**{**switch_dict, **{"_id": dpid, "updated_at": utc_now}})
+        updated = self.db.switches.find_one_and_update(
+            {"_id": dpid},
+            {
+                "$set": model.dict(exclude={"inserted_at"}),
+                "$setOnInsert": {"inserted_at": utc_now},
+            },
+            return_document=ReturnDocument.AFTER,
+            upsert=True,
+        )
+        return updated
+```
+
+The following Python package and module structure is encouraged, that way NApps that are using this architecture it's evident where things are and how they are generally supposed to be evolved, facilitating for new developers too. On each package, you can start with a `__init__.py` module, as illustrated below and put the classes there, and if necessary you can create new modules under a specific package and move them. `api/models` don't necessarily need to be `pydantic` ones, you can also leverage `MongoDB` projections or use existing models that have already been implemented and tested:
```

Comment:
```suggestion
The following Python package and module structure is encouraged, that way for NApps using this architecture it's clear where things are and how they are supposed to evolve, facilitating for new developers too. On each package, you can start with a `__init__.py` module, as illustrated below and put the classes there, and if necessary you can create new modules under a specific package and move them. `api/models` don't necessarily need to be `pydantic` ones, you can also leverage `MongoDB` projections or use existing models that have already been implemented and tested:
```

Reply:
Thanks for this suggestion @ajoaoff, just committed it.